src/lattice_reaction_system_diffusion.jl

```diff
@@ -0,0 +1,461 @@
+### Diffusion Reaction Structure. ###
+
+# Implements the diffusionparameter metadata field.
+struct DiffusionParameter end
+Symbolics.option_to_metadata_type(::Val{:diffusionparameter}) = DiffusionParameter
+
+isdiffusionparameter(x::Num, args...) = isdiffusionparameter(Symbolics.unwrap(x), args...)
+function isdiffusionparameter(x, default = false)
+    p = Symbolics.getparent(x, nothing)
+    p === nothing || (x = p)
+    Symbolics.getmetadata(x, DiffusionParameter, default)
+end
+
+# Abstract spatial reaction structures.
+abstract type AbstractSpatialReaction end
+
+# A diffusion reaction. These are simple to hanlde, and should cover most types of spatial reactions.
+# Currently only permit constant rates.
+struct DiffusionReaction <: AbstractSpatialReaction
+    """The rate function (excluding mass action terms). Currently only constants supported"""
+    rate::Num
+    """The species that is subject to difusion."""
+    species::Num
+    """A symbol representation of the species that is subject to difusion."""
+    species_sym::Symbol    # Required for identification in certain cases.
+
+    # Creates a diffusion reaction.
+    function DiffusionReaction(rate::Num, species::Num)
+        new(rate, species, ModelingToolkit.getname(species))
+    end
+    function DiffusionReaction(rate::Number, species::Num)
+        new(Num(rate), species, ModelingToolkit.getname(species))
+    end
+    function DiffusionReaction(rate::Symbol, species::Num)
+        new(Symbolics.variable(rate), species, ModelingToolkit.getname(species))
+    end
+    function DiffusionReaction(rate::Num, species::Symbol)
+        new(rate, Symbolics.variable(species), species)
+    end
+    function DiffusionReaction(rate::Number, species::Symbol)
+        new(Num(rate), Symbolics.variable(species), species)
+    end
+    function DiffusionReaction(rate::Symbol, species::Symbol)
+        new(Symbolics.variable(rate), Symbolics.variable(species), species)
+    end
+end
+# Creates a vector of DiffusionReactions.
+function diffusion_reactions(diffusion_reactions)
+    [DiffusionReaction(dr[1], dr[2]) for dr in diffusion_reactions]
+end
+# Gets the parameters in a diffusion reaction.
+ModelingToolkit.parameters(dr::DiffusionReaction) = Symbolics.get_variables(dr.rate)
+
+### Lattice Reaction Network Structure ###
+# Desribes a spatial reaction network over a graph.
+struct LatticeReactionSystem # <: MT.AbstractTimeDependentSystem # Adding this part messes up show, disabling me from creating LRSs
```

Comment:
If we want to make it an `AbstractTimeDependentSystem`, which seems sensible, it then needs to define versions of the MTK API functions. I don't think they are fully documented anywhere, so probably we'd need to just look at what `ReactionSystem` overloads and mimic that.

Reply:
Why not make this parametric so the fields are all concretely typed?

Reply:
Before today I didn't know there was a practical difference between concrete and non-concrete types, nor that there were some performance advantages to using parametric types. But if phrasing it that way improves performance then we should do it like that, yes.

Reply:
Yes, generally you want all the fields within a struct to have a concrete type by either explicitly specifying the type or using parametric types. Using abstract / non-concrete types like AbstractVector or even Vector (as opposed to Vector{Float64}) can have performance implications.

We have to make a few fields non-concrete in ReactionSystem as not all the symbolic stuff is concrete and/or they store mixed types. But where possible it should be avoided.

Reply:
I'd suggest not worring about matching the MTK system API in this PR, we can followup on that. But making it parametric / concrete would be good.

Reply:
Sounds good

Reply:
Things are now concretely typed, (with `LatticeReactionSystems` becoming `<:` of `MT.AbstractTimeDependentSystem` something for the future).